File: packages/mda-config/typescript/src/integrity.ts
/**
 * MDA §08 — integrity canonicalization, hashing, and verification.
 *
 * Canonical bytes recipe (§08-3):
 *   1. Strip top-level `integrity` and `signatures[]` from the frontmatter.
 *   2. Convert the stripped frontmatter to JCS (RFC 8785).
 *   3. Normalize the body string per §08-3.3 (LF endings already handled by
 *      §02-1.1, then strip trailing spaces/tabs per line, then ensure exactly
 *      one terminating "\n" UNLESS the body is empty).
 *   4. Concatenate as: b"---\n" + jcs + b"\n---\n" + normalized_body.
 *
 * Multi-file artifacts (§08-3.4) are out of scope for v1.0 (single-file source-
 * mode loader); the helper accepts only one file.
 */

import { canonify } from "@truestamp/canonify";
import { createHash } from "node:crypto";
import { ErrorCategory, MdaConfigError } from "./errors.js";

/** Shape of the optional top-level `integrity` field (§08-2). */
export interface IntegrityField {
  algorithm: "sha256" | "sha384" | "sha512";
  digest: string;
}

/** §08-3.1 — strip `integrity` and `signatures` from a shallow frontmatter copy. */
function stripSecurityFields(
  frontmatter: Record<string, unknown>,
): Record<string, unknown> {
  const copy = { ...frontmatter };
  delete copy.integrity;
  delete copy.signatures;
  return copy;
}

/** §08-3.3 — normalize the body string for digest computation. */
export function normalizeBody(bodyStr: string): string {
  if (bodyStr === "") return "";
  // §02-1.1 step 3 already handled CRLF/CR → LF; restate per §08-3.3 step 1.
  // §08-3.3 step 2 — strip trailing spaces/tabs from each line.
  const lines = bodyStr.split("\n");
  const stripped = lines.map((l) => l.replace(/[ \t]+$/u, ""));
  // §08-3.3 step 3 — exactly one terminating "\n".
  // After split("\n"), a string ending in "\n" produces an empty last segment.
  // Drop trailing empty segments and re-append exactly one newline.
  while (stripped.length > 0 && stripped[stripped.length - 1] === "") {
    stripped.pop();
  }
  if (stripped.length === 0) return "";
  return stripped.join("\n") + "\n";
}

/** §08-3 — assemble the canonical artifact bytes for a single-file source. */
export function canonicalizeArtifact(
  frontmatter: Record<string, unknown>,
  bodyStr: string,
): Uint8Array {
  // §08-3.1 + §08-3.2 — strip security fields, JCS-canonicalize.
  const stripped = stripSecurityFields(frontmatter);
  const jcs = canonify(stripped);
  // §08-3.3 — normalize body bytes.
  const normalizedBody = normalizeBody(bodyStr);
  // §08-3.3 — concatenation recipe.
  const head = "---\n" + jcs + "\n---\n";
  return new TextEncoder().encode(head + normalizedBody);
}

/** §08-2 — parse a self-describing digest "<algorithm>:<lowercase-hex>". */
export function parseDigest(digest: string): { algorithm: string; hex: string } {
  const idx = digest.indexOf(":");
  if (idx <= 0) {
    throw new MdaConfigError(
      ErrorCategory.SchemaViolation,
      "integrity.digest is not in '<algorithm>:<hex>' form",
      { digest },
    );
  }
  return { algorithm: digest.slice(0, idx), hex: digest.slice(idx + 1) };
}

/** §08-3.5 — hash canonical bytes with the declared algorithm. */
export function hashCanonical(
  canonicalBytes: Uint8Array,
  algorithm: "sha256" | "sha384" | "sha512",
): string {
  return createHash(algorithm).update(canonicalBytes).digest("hex");
}

/** §08-4 — verify the declared `integrity.digest` matches the recomputed hash. */
export function verifyIntegrity(
  frontmatter: Record<string, unknown>,
  bodyStr: string,
  integrity: IntegrityField,
): void {
  const parsed = parseDigest(integrity.digest);
  if (parsed.algorithm !== integrity.algorithm) {
    throw new MdaConfigError(
      ErrorCategory.SchemaViolation,
      "integrity.digest prefix does not match integrity.algorithm",
      { algorithm: integrity.algorithm, digestPrefix: parsed.algorithm },
    );
  }
  const canonical = canonicalizeArtifact(frontmatter, bodyStr);
  const computed = hashCanonical(canonical, integrity.algorithm);
  if (computed !== parsed.hex) {
    throw new MdaConfigError(
      ErrorCategory.IntegrityMismatch,
      "computed digest does not match integrity.digest",
      { expected: parsed.hex, computed, algorithm: integrity.algorithm },
    );
  }
}

File: packages/mda-config/rust/src/integrity.rs
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use sha2::{Digest, Sha256, Sha384, Sha512};

use crate::errors::{ErrorCategory, MdaConfigError, Result};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum HashAlgorithm {
    Sha256,
    Sha384,
    Sha512,
}

impl HashAlgorithm {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Sha256 => "sha256",
            Self::Sha384 => "sha384",
            Self::Sha512 => "sha512",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
pub struct IntegrityField {
    pub algorithm: HashAlgorithm,
    pub digest: String,
}

pub fn normalize_body(body_str: &str) -> String {
    if body_str.is_empty() {
        return String::new();
    }
    let mut stripped: Vec<String> = body_str
        .split('\n')
        .map(|line| line.trim_end_matches([' ', '\t']).to_string())
        .collect();
    while stripped.last().is_some_and(|line| line.is_empty()) {
        stripped.pop();
    }
    if stripped.is_empty() {
        String::new()
    } else {
        stripped.join("\n") + "\n"
    }
}

pub fn canonicalize_artifact(frontmatter: &Value, body_str: &str) -> Result<Vec<u8>> {
    let object = frontmatter.as_object().ok_or_else(|| {
        MdaConfigError::new(
            ErrorCategory::SchemaViolation,
            "frontmatter MUST be a JSON object",
        )
    })?;
    let mut stripped = object.clone();
    stripped.remove("integrity");
    stripped.remove("signatures");
    let canonical = canonical_json(&Value::Object(stripped))?;
    let head = format!("---\n{canonical}\n---\n");
    Ok([head.as_bytes(), normalize_body(body_str).as_bytes()].concat())
}

pub fn parse_digest(digest: &str) -> Result<(&str, &str)> {
    let Some((algorithm, hex)) = digest.split_once(':') else {
        return Err(MdaConfigError::with_details(
            ErrorCategory::SchemaViolation,
            "integrity.digest is not in '<algorithm>:<hex>' form",
            serde_json::json!({ "digest": digest }),
        ));
    };
    if algorithm.is_empty() {
        return Err(MdaConfigError::with_details(
            ErrorCategory::SchemaViolation,
            "integrity.digest is not in '<algorithm>:<hex>' form",
            serde_json::json!({ "digest": digest }),
        ));
    }
    Ok((algorithm, hex))
}

pub fn hash_canonical(canonical_bytes: &[u8], algorithm: HashAlgorithm) -> String {
    match algorithm {
        HashAlgorithm::Sha256 => hex_lower(Sha256::digest(canonical_bytes).as_slice()),
        HashAlgorithm::Sha384 => hex_lower(Sha384::digest(canonical_bytes).as_slice()),
        HashAlgorithm::Sha512 => hex_lower(Sha512::digest(canonical_bytes).as_slice()),
    }
}

pub fn verify_integrity(
    frontmatter: &Value,
    body_str: &str,
    integrity: &IntegrityField,
) -> Result<()> {
    let (algorithm, expected_hex) = parse_digest(&integrity.digest)?;
    if algorithm != integrity.algorithm.as_str() {
        return Err(MdaConfigError::with_details(
            ErrorCategory::SchemaViolation,
            "integrity.digest prefix does not match integrity.algorithm",
            serde_json::json!({
                "algorithm": integrity.algorithm.as_str(),
                "digestPrefix": algorithm,
            }),
        ));
    }
    let canonical = canonicalize_artifact(frontmatter, body_str)?;
    let computed = hash_canonical(&canonical, integrity.algorithm);
    if computed != expected_hex {
        return Err(MdaConfigError::with_details(
            ErrorCategory::IntegrityMismatch,
            "computed digest does not match integrity.digest",
            serde_json::json!({
                "expected": expected_hex,
                "computed": computed,
                "algorithm": integrity.algorithm.as_str(),
            }),
        ));
    }
    Ok(())
}

pub(crate) fn canonical_json(value: &Value) -> Result<String> {
    match value {
        Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_) => {
            serde_json::to_string(value).map_err(|cause| {
                MdaConfigError::with_details(
                    ErrorCategory::SchemaViolation,
                    "JSON canonicalization failed",
                    serde_json::json!({ "cause": cause.to_string() }),
                )
            })
        }
        Value::Array(items) => {
            let mut out = String::from("[");
            for (idx, item) in items.iter().enumerate() {
                if idx > 0 {
                    out.push(',');
                }
                out.push_str(&canonical_json(item)?);
            }
            out.push(']');
            Ok(out)
        }
        Value::Object(map) => canonical_object(map),
    }
}

fn canonical_object(map: &Map<String, Value>) -> Result<String> {
    let mut keys: Vec<&String> = map.keys().collect();
    keys.sort();
    let mut out = String::from("{");
    for (idx, key) in keys.iter().enumerate() {
        if idx > 0 {
            out.push(',');
        }
        out.push_str(&serde_json::to_string(key).map_err(|cause| {
            MdaConfigError::with_details(
                ErrorCategory::SchemaViolation,
                "JSON canonicalization failed",
                serde_json::json!({ "cause": cause.to_string() }),
            )
        })?);
        out.push(':');
        out.push_str(&canonical_json(&map[*key])?);
    }
    out.push('}');
    Ok(out)
}

fn hex_lower(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut out = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        out.push(HEX[(byte >> 4) as usize] as char);
        out.push(HEX[(byte & 0x0f) as usize] as char);
    }
    out
}
